sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala

```diff
@@ -407,260 +411,237 @@ class GpuSequenceMeta(
+          val stopCol = stopGpuCol.getBase
+          val steps = stepGpuColOpt.map(_.getBase.incRefCount())
+              .getOrElse(defaultSteps(startCol, stopCol))
+          closeOnExcept(steps) { _ =>
+            (computeSequenceSizes(startCol, stopCol, steps), steps)
           }
-          (sizeWithNegative, tmpDiffHasZero)
         }
 
-        val tmpSize = closeOnExcept(diffHasZero) { _ =>
-          // tmpSize = if(sizeWithNegative < 0) 0 else sizeWithNegative
-          withResource(sizeWithNegative) { _ =>
-            withResource(sizeWithNegative.greaterOrEqualTo(zero)) { pred =>
-              pred.ifElse(sizeWithNegative, zero)
-            }
+        // 2 Generate the sequence
+        //
+        // cudf 'sequence' requires 'step' has the same type with 'start'.
+        // And the step type may differ due to the default steps.
+        val castedStepCol = withResource(stepCol) { _ =>
+          closeOnExcept(sizeCol) { _ =>
+            stepCol.castTo(startCol.getType)
           }
         }
-
-        // when start==stop && step==0, size will be 0.
-        // but we should change size to 1
-        withResource(tmpSize) { tmpSize =>
-          withResource(diffHasZero) { diffHasZero =>
-            step match {
-              case stepScalar: Scalar =>
-                withResource(ColumnVector.fromScalar(stepScalar, rows)) { stepV =>
-                  withResource(stepV.equalTo(zero)) { stepHasZero =>
-                    withResource(diffHasZero.and(stepHasZero)) { predWithZero =>
-                      predWithZero.ifElse(one, tmpSize)
-                    }
-                  }
-                }
-              case _ =>
-                withResource(step.equalTo(zero)) { stepHasZero =>
-                  withResource(diffHasZero.and(stepHasZero)) { predWithZero =>
-                    predWithZero.ifElse(one, tmpSize)
-                  }
-                }
-            }
-          }
+        withResource(Seq(sizeCol, castedStepCol)) { _ =>
+          GpuColumnVector.from(genSequence(startCol, sizeCol, castedStepCol), dataType)
         }
       }
     }
   }
 
-  override def doColumnar(
-      start: GpuColumnVector,
-      stop: GpuColumnVector,
-      step: GpuColumnVector): ColumnVector = {
-    withResource(calculateSize(start.getBase, stop.getBase, step.getBase, start.getRowCount.toInt,
-        start.dataType())) { size =>
-      ColumnVector.sequence(start.getBase, size, step.getBase)
-    }
-  }
-
-  override def doColumnar(
-      start: GpuScalar,
-      stop: GpuColumnVector,
-      step: GpuColumnVector): ColumnVector = {
-    withResource(calculateSize(start.getBase, stop.getBase, step.getBase, stop.getRowCount.toInt,
-        start.dataType)) { size =>
-      withResource(ColumnVector.fromScalar(start.getBase, stop.getRowCount.toInt)) { startV =>
-        ColumnVector.sequence(startV, size, step.getBase)
+  @transient
+  private lazy val defaultSteps: (ColumnView, ColumnView) => ColumnVector =
```

Comment:
This cannot be a `lazy val` and we cannot in general use `lazy val` on GPU resources.  If it is needed twice, the first time it will create a `ColumnVector` instance and then return it, and the caller will eventually close it.  Then the second time we will _not_ create a new instance, because it's a `lazy val`, and we'll end up returning an instance that corresponds to a closed column vector.  This needs to be a `def` not a `val`, as we need to generate the column each time since it will be closed each time.

Reply:
This `defaultSteps` is just a function and its type is `(ColumnView, ColumnView) => ColumnVector`, not a GPU resource.

OK, the name is confusable, updated. Sorry for that.

Reply:
Apologies for misreading this initially, but I'm still not seeing a compelling reason to use a `lazy val` here instead of a more readable `def` approach.  It seems like this is all to save a single match on `dataType`, otherwise it's just like any other def (but harder to read).